File: client-java/controller/src/main/java/org/evomaster/clientJava/controller/EmbeddedSutController.java
package org.evomaster.clientJava.controller;

import org.evomaster.clientJava.controller.internal.SutController;
import org.evomaster.clientJava.instrumentation.AdditionalInfo;
import org.evomaster.clientJava.instrumentation.InstrumentationController;
import org.evomaster.clientJava.instrumentation.InstrumentingAgent;
import org.evomaster.clientJava.instrumentation.TargetInfo;
import org.evomaster.clientJava.instrumentation.staticstate.ExecutionTracer;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;


public abstract class EmbeddedSutController extends SutController {

    @Override
    public final boolean isInstrumentationActivated() {
        return InstrumentingAgent.isActive();
    }

    @Override
    public final void newSearch(){
        InstrumentationController.resetForNewSearch();
    }

    @Override
    public final void newTest(){
        InstrumentationController.resetForNewTest();
        resetExtraHeuristics();
    }

    @Override
    public final List<TargetInfo> getTargetInfos(Collection<Integer> ids){
        return InstrumentationController.getTargetInfos(ids);
    }

    @Override
    public final List<AdditionalInfo> getAdditionalInfoList(){
        return InstrumentationController.getAdditionalInfoList();
    }

    @Override
    public final void newAction(int actionIndex){
        ExecutionTracer.setActionIndex(actionIndex);
        resetExtraHeuristics();
    }
}

File: client-dotnet/src/EvoMaster.Controller/EmbeddedSutController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using EvoMaster.Controller.Api;
using EvoMaster.Instrumentation;
using EvoMaster.Instrumentation.StaticState;
using Microsoft.Extensions.Configuration;
using Action = EvoMaster.Instrumentation.Action;

namespace EvoMaster.Controller {
    public abstract class EmbeddedSutController : SutController {
        public sealed override UnitsInfoDto GetUnitsInfoDto() {
            // return GetUnitsInfoDto(UnitsInfoRecorder.GetInstance());
            return GetUnitsInfoDto(UnitsInfoRecorder.GetInstance());
        }

        public sealed override bool IsInstrumentationActivated() => false;

        public sealed override void NewActionSpecificHandler(ActionDto dto) {
            ExecutionTracer.SetAction(new Action(dto.Index, dto.InputVariables));
        }

        public override IList<TargetInfo> GetTargetInfos(IEnumerable<int> ids) {
            return InstrumentationController.GetTargetInfos(ids);
        }

        public override IList<AdditionalInfo> GetAdditionalInfoList() {
            return InstrumentationController.GetAdditionalInfoList();
        }

        public sealed override void NewSearch() {
            InstrumentationController.ResetForNewSearch();
        }

        public sealed override void NewTestSpecificHandler() {
            InstrumentationController.ResetForNewTest();
        }

        public override void SetKillSwitch(bool b) => ExecutionTracer.SetKillSwitch(b);

        /// <summary>
        /// This method checks whether the SUT is listening on the port number
        /// </summary>
        /// <param name="port">The port number on the localhost</param>
        /// <param name="timeout">The amount of time in seconds the driver should give up if the SUT did not start </param>
        protected static void WaitUntilSutIsRunning(int port, int timeout = 30) {
            var task = Task.Run(() => {
                using var tcpClient = new TcpClient();
                while (true) {
                    try {
                        tcpClient.Connect("127.0.0.1", port);
                        break;
                    }
                    catch (Exception) {
                        Thread.Sleep(50);
                    }
                }
            });

            if (task.Wait(TimeSpan.FromSeconds(timeout)))
                return;

            throw new TimeoutException($"The SUT didn't start within {timeout} seconds on port {port}.");
        }

        protected static IConfiguration GetConfiguration(string settingFileName) {
            var config = new ConfigurationBuilder()
                .AddJsonFile(settingFileName)
                .Build();
            return config;
        }
    }
}